common/src/chain/transaction/signature/inputsig/classical_multisig/authorize_classical_multisig.rs

```diff
@@ -0,0 +1,788 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use crypto::key::Signature;
+use serialization::{Decode, Encode};
+
+use crate::{
+    address::pubkeyhash::PublicKeyHash,
+    chain::{
+        classic_multisig::{ClassicMultisigChallenge, ClassicMultisigChallengeError},
+        signature::{
+            inputsig::classical_multisig::multisig_partial_signature::PartiallySignedMultisigChallenge,
+            TransactionSigError,
+        },
+        ChainConfig,
+    },
+    primitives::H256,
+};
+
+use super::multisig_partial_signature::PartiallySignedMultisigStructureError;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ClassicalMultisigCompletion {
+    Complete(AuthorizedClassicalMultisigSpend),
+    Incomplete(AuthorizedClassicalMultisigSpend),
+}
+
+impl ClassicalMultisigCompletion {
+    pub fn take(self) -> AuthorizedClassicalMultisigSpend {
+        match self {
+            Self::Complete(spend) => spend,
+            Self::Incomplete(spend) => spend,
+        }
+    }
+
+    pub fn is_complete(&self) -> bool {
+        matches!(self, Self::Complete(_))
+    }
+}
+
+/// A witness that represents the authorization to spend a classical multisig output.
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub struct AuthorizedClassicalMultisigSpend {
+    signatures: BTreeMap<u8, Signature>,
+    challenge: ClassicMultisigChallenge,
+}
+
+impl AuthorizedClassicalMultisigSpend {
```

Comment:
@TheQuantumPhysicist this is where I'd propose beefing the commenting (or the test but you seemed more pro-commenting and idgaf about which we do as they have the same outcome to me). I'd propose adding some comments about how signatures are handled here, specifically the map/index stuff. It's clear if you spend a little time looking at the code but not necessarily when running through it.